src/pipeline/src/etl/processor/vrl.rs

```diff
@@ -0,0 +1,319 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+
+use chrono_tz::Tz;
+use snafu::{OptionExt, ResultExt};
+use vrl::compiler::runtime::Runtime;
+use vrl::compiler::{compile, Program, TargetValue};
+use vrl::diagnostic::Formatter;
+use vrl::prelude::{Bytes, NotNan, TimeZone};
+use vrl::value::{KeyString, Kind, Secrets, Value as VrlValue};
+
+use crate::error::{
+    BytesToUtf8Snafu, CompileVrlSnafu, Error, ExecuteVrlSnafu, FloatNaNSnafu,
+    InvalidTimestampSnafu, KeyMustBeStringSnafu, Result, VrlRegexValueSnafu, VrlReturnValueSnafu,
+};
+use crate::etl::processor::yaml_string;
+use crate::{PipelineMap, Value as PipelineValue};
+
+pub(crate) const PROCESSOR_VRL: &str = "vrl";
+const SOURCE: &str = "source";
+
+#[derive(Debug)]
+pub struct VrlProcessor {
+    source: String,
+    program: Program,
+}
+
+impl VrlProcessor {
+    pub fn new(source: String) -> Result<Self> {
+        let fns = vrl::stdlib::all();
+
+        let compile_result = compile(&source, &fns).map_err(|e| {
+            CompileVrlSnafu {
+                msg: Formatter::new(&source, e).to_string(),
+            }
+            .build()
+        })?;
+
+        let program = compile_result.program;
+
+        // check if the return value is have regex
+        let result_def = program.final_type_info().result;
+        let kind = result_def.kind();
+        if !kind.is_object() {
+            return VrlReturnValueSnafu.fail();
+        }
+        check_regex_output(kind)?;
+
+        Ok(Self { source, program })
+    }
+
+    pub fn resolve(&self, m: PipelineMap) -> Result<PipelineValue> {
+        let pipeline_vrl = m
+            .into_iter()
+            .map(|(k, v)| pipeline_value_to_vrl_value(v).map(|v| (KeyString::from(k), v)))
+            .collect::<Result<BTreeMap<_, _>>>()?;
+
+        let mut target = TargetValue {
+            value: VrlValue::Object(pipeline_vrl),
+            metadata: VrlValue::Object(BTreeMap::new()),
+            secrets: Secrets::default(),
+        };
+
+        let timezone = TimeZone::Named(Tz::UTC);
+        let mut runtime = Runtime::default();
+        let re = runtime
+            .resolve(&mut target, &self.program, &timezone)
+            .map_err(|e| {
+                ExecuteVrlSnafu {
+                    msg: e.get_expression_error().to_string(),
+                }
+                .build()
+            })?;
+
+        vrl_value_to_pipeline_value(re)
+    }
+}
+
+impl TryFrom<&yaml_rust::yaml::Hash> for VrlProcessor {
```

Comment:
We can make this `source` key optional for better ergonomic. If the value of `vrl` is a string, just use it as source.

Reply:
Due to implementation details, we have to keep the source key(for parsing the YAML of each processor, the string literal is converted into a hash map before it goes into the processor's independent parsing logic). Also, we might add other keys like config options later for compiling the vrl source code.